### PR TITLE
Add type for gpio::AnyPin

### PR DESCRIPTION
**Pull request by Volkalex28**
This makes it possible to safely implement the InputPin and OutputPin traits for AnyPin. Now you can convert any pin to AnyPin with the appropriate type and use it in other library modules

Added:
- Peripheral implementation for AnyPin
- Implementation of Pin for AnyPin
- Implementation of OutputPin for AnyPin with type IsOutputPin
- Implementation of InputPin for AnyPin with type IsInputPin
- Upgrade types for AnyPin (for example InputOutputAnalogPinType -> InputOutputPinType)
- Implementation of From<Gpio> for AnyPin with the appropriate type

Changed:
- The Gpio::degrage method returns AnyPin with the appropriate type

A small example where this can be useful:
```rust
#![no_std]
#![no_main]

use esp32s3_hal::{clock::ClockControl, gpio::IO, i2c::I2C, peripherals::Peripherals, prelude::*};
use esp_backtrace as _;
use esp_println::println;

enum BoardRevision {
    Dev,
    Release
}

// Take data about the device on which the program is running. 
// This is not known at the compilation stage, but is extracted at runtime,
// for example from the device serial number
fn get_board_revision() -> BoardRevision {
    //...
}

#[entry]
fn main() -> ! {
    let peripherals = Peripherals::take();
    let system = peripherals.SYSTEM.split();
    let clocks = ClockControl::boot_defaults(system.clock_control).freeze();

    let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
    
    // Select pins depending on BoardRevision
    let (sda, scl) = match get_board_revision() {
        BoardRevision::Dev => (io.pins.gpio43.degrade(), io.pins.gpio44.degrade()),
        BoardRevision::Release=>(io.pins.gpio41.degrade(), io.pins.gpio38.degrade()),
    }

    // Create a new peripheral object with the described wiring
    // and standard I2C clock speed
    let mut i2c = I2C::new(peripherals.I2C0, sda, scl, 100u32.kHz(), &clocks);

    loop {
        let mut data = [0u8; 22];
        i2c.write_read(0x77, &[0xaa], &mut data).ok();

        println!("{:02x?}", data);
    }
}
```

### CODE REVIEW

**Comment by bjoernQ** [comment]
Thanks! This is very useful

Maybe the `.degrade().degrade()` looks a bit weird in the example - probably `.degrade().into()` would look better

Other than that, this looks really good to me

**Comment by jessebraham** [comment]
Not a fan of the API honestly (`led3.degrade().into()` means absolutely nothing to me looking at the examples, I can only imagine how confusing this will be for beginners), but I also don't really have any better suggestions so no point blocking it I guess.

**Comment by jessebraham** [comment]
@Volkalex28 if you're able to rebase this, then I think we're good to merge.

**Comment by Volkalex28** [comment]
Sorry for delay. I'll rebase this tomorrow.

**Comment by Volkalex28** [comment]
@bjoernQ Please check it after rebase

File: CHANGELOG.md
Comment (bjoernQ) [inline CHANGELOG.md:104]: this needs to be moved to the unreleased section and the PR number added

**Review by bjoernQ — APPROVED** [review]
LGTM - needs the changelog fixed before merging

**Comment by bjoernQ** [comment]
Thanks! We can merge this after fixing the changelog

**Comment by Volkalex28** [comment]
> Thanks! We can merge this after fixing the changelog

Let me know when you can merge this or just do it 😉

**Comment by bjoernQ** [comment]
Sorry the changelog entry should have been added to Unreleased/Added but I probably wasn't clear about it. Going to merge it as is and I'll fix the changelog afterwards. Thanks again

**Pull request merged**
